### PR TITLE
feat: add herewallet

### PR DESCRIPTION
**Pull request by AZbang**
# Description

[Here Wallet](https://herewallet.app/) is **mobile** waller for Near Protocol

You can read about first steps in this [article](https://docs.herewallet.app/app/first-step)

| Web bridge  | Mobile App |
| ------------- | ------------- | 
| image  |  image | 

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by kujtimprenkuSQA** [comment]
Hey, @AZbang thank you for submitting this PR.

Could you please change the base branch from `main` to `dev` it is explained here:
https://github.com/near/wallet-selector/blob/main/CONTRIBUTING.md#branching

Can you also explain how can we use this wallet with Wallet Selector?

After clicking the Here Wallet option I get redirected to this [website](https://web.testnet.herewallet.app/login/?success_url=http%3A%2F%2Flocalhost%3A4200%2F&failure_url=http%3A%2F%2Flocalhost%3A4200%2F&contract_id=guest-book.testnet&public_key=ed25519%3AB3E7CQEhYf3H5Fzif8uu3MYpoQgo3FEUZfSihH6CjaD4) in the right side it says scan QR Code but in the HERE app in IOS which seems to be still work in progress I can not find any way to scan the QR code.

What I have tried in the app:
Import `kujtimprenku.testnet` via Seed Phrase
Create a new account through the app's UI: `kujtim.near`

Can we use the guest book example as it is, does the app (Here Wallet) support testnet?

**Comment by AZbang** [comment]
@kujtimprenkuSQA Hello! Yes, Here Wallet supports testnet.

If you select here wallet in the guest book example app and try to send the message, you will be redirected to the site
web.testnet.herewallet.app for sign transactions. Also on this site you can download testnet version of Here Wallet (https://testflight.apple.com/join/LwvGXAK8). After installation, you can create a new account or export an existing one.

To send the transaction -- scan the qr code through your iPhone's native camera app, the camera will prompt you to open the Here Testnet app (for mainnet qr codes, you will be prompted to open the mainnet version of the app, available here https://testflight.apple.com/join/osMqBAZ9)

There is currently no way to scan the qr code in the app itself, we are working on it, however the qr is just a link with a custom url scheme, so you can use any mobile scanner, we recommend using the native iPhone camera app)

For the mainnet version, the usage is absolutely identical, you will just be redirected to web.herewallet.app to interact with your wallet

**Comment by Neafiol** [comment]
I confirm that the wallet meets the criteria from https://github.com/near/wallet-selector/blob/main/CONTRIBUTING.md#branching

**Comment by AZbang** [comment]
@kujtimprenkuSQA  i changed the base branch from main to dev, thank you!

File: packages/here-wallet/src/lib/utils.ts
Comment (kujtimprenkuSQA) [inline packages/here-wallet/src/lib/utils.ts:76]: Can we avoid using a `Promise.all()` here?

This is to potentially avoid an issue that was reported here in Wallet Selector in the past: this avoids nonce increment issues and a concept called “transaction dependency relationship“ where the order of the transaction matters.

You can use a regular for loop to create these transactions in series.

File: packages/here-wallet/src/lib/here-wallet.ts
Comment (kujtimprenkuSQA) [inline packages/here-wallet/src/lib/here-wallet.ts:134]: Can you please add the icon in PNG format in the assets folder and provide it in a similar way as it is in other wallets, and update the readme to keep things consistent?

File: packages/here-wallet/package.json
Comment (kujtimprenkuSQA) [inline packages/here-wallet/package.json:3]: Can you update the version and set it to the same version as the other packages?

Also for packages that depend on `near-api-js` we need to provide it as a peerDependency the same way we have done for `@near-wallet-selector/my-near-wallet` and other packages see here:

https://github.com/near/wallet-selector/blob/main/packages/my-near-wallet/package.json#L4-L6

**Review by kujtimprenkuSQA — COMMENTED** [review]
Hey, thank you for switching the base branch to `dev` and providing the link for `testnet` to test the wallet. Just a few comments and questions from me.

File: packages/here-wallet/src/lib/here-wallet.ts
Comment (AZbang) [reply packages/here-wallet/src/lib/here-wallet.ts:19]: These are unnecessary methods but that make it possible to get the correct wallet balance.
Here Wallet users often transfer all their near tokens into a here wallet smart contract for liquid staking. Therefore, the classic way to get a balance does not return the full amount of tokens available for spending. getAvailableBalance solve this problem and these methods are well documented in the readme. It seems to me that each wallet has the right to add some of its own custom methods if this will help library users work more efficiently, am I wrong?

```ts
const isHereWallet = (w: Wallet): w is HereWallet =>
  w.id === "here-wallet";
  
if (isHereWallet(wallet)) {
  wallet.getAvailableBalance(); // correct typings
}
```

File: packages/here-wallet/src/lib/utils.ts
Comment (AZbang) [reply packages/here-wallet/src/lib/utils.ts:76]: But Promise.all doesn't change the order of the array elements:
https://stackoverflow.com/questions/28066429/promise-all-order-of-resolved-values

My-near-wallet has a similar code in the package.

File: packages/here-wallet/src/lib/here-wallet.ts
Comment (AZbang) [reply packages/here-wallet/src/lib/here-wallet.ts]: I've done

**Comment by kujtimprenkuSQA** [comment]
Thank you for resolving those comments.

When I send one message from guest book example and I attach a donation/deposit (NEAR) to the message on the first try it works correctly but when I immediately try to send a similar message and attach a deposit I see this screen:

request-uri

**Comment by AZbang** [comment]
@kujtimprenkuSQA 
It seems we have increased the request length limit, please try again. Thank you!
If the problem persists, please send a link to your request to web.testnet.nearwallet.app, so we can quickly understand what the problem is.

However, there are browser limits that we cannot overcome, for example, if we send a very long request to MyNearWallet, we will get a similar error about the length of the request.

**Comment by kujtimprenkuSQA** [comment]
Hey, @AZbang when I send a transaction with a deposit for the first time and I approve it on my mobile phone after I get redirected from https://web.testnet.herewallet.app/ these params are being added to the guest book example:

http://localhost:4200/?account_id=kujtimprenku.testnet&meta=undefined&all_keys=ed25519%3A8Bu6bMh....

Please note: I did not paste the complete URL because it is 10207 characters long.

I think the reason it is failing the second time you try to send a transaction with a deposit is that https://web.testnet.herewallet.app/  tries to construct a `callbackUrl` to redirect to localhost and includes also the query params which are 10k+ characters.

I personally can not pinpoint the issue here but the only way to send a second transaction with a deposit is to manually remove the query params **?account_id=kujtimprenku.testnet&meta=undefined&all_keys=ed25519%3A8Bu6bMh...**. from localhost:4200  then reload the page after that send the transaction which is not ideal at all.

To reproduce it in the guest book example:
 - Send 1 message with a deposit and approve it.
 - After successfully being redirected to the guest book example try sending another message with a deposit.

**Comment by AZbang** [comment]
@kujtimprenkuSQA Thanks for the good problem description! We fixed this bug, please check again

**Comment by AmmarHumackicSQA** [comment]
Hey @AZbang as a reminder, the security statement needs to be shown either on your GitHub repo or the website. You can check it [here](https://github.com/near/wallet-selector/blob/main/CONTRIBUTING.md#:~:text=GitHub%20account%20or%20Website) again. Thanks!

**Comment by Neafiol** [comment]
we stated it in the company profile

https://github.com/here-wallet
image

**Comment by AmmarHumackicSQA** [comment]
> we stated it in the company profile
> 
> https://github.com/here-wallet image

Thank you @Neafiol !

**Comment by Neafiol** [comment]
We are already approved in Apple Store, we are planning release on the 28th of September. Now we are waiting for the addition to the Wallet Select and the readiness of our partners

**Comment by kujtimprenkuSQA** [comment]
>We are already approved in Apple Store, we are planning release on the 28th of September. Now we are waiting for the addition to the Wallet Select and the readiness of our partners

Thanks for letting us know.
**The integration with Wallet Selector works as expected** 👍 .

It seems like there are some conflicts with the current `dev` branch, can you please have a look at them, as soon as those are fixed we can merge this PR to `dev`.

**Comment by AZbang** [comment]
@kujtimprenkuSQA i resolved it. Thank you

**Review by kujtimprenkuSQA — APPROVED** [review]

**Comment by AmmarHumackicSQA** [comment]
> We are already approved in Apple Store, we are planning release on the 28th of September. Now we are waiting for the addition to the Wallet Select and the readiness of our partners

Hi @Neafiol @AZbang the PR has been approved, when can we expect the app to be published on App Store so we can merge this and include it in the next release of Wallet Selector? Thanks.

**Pull request merged**
